docs/guides/implementation-guides/how-to-implement-dependency-injection.md

```diff
@@ -7,9 +7,120 @@ import MvvmArchitectureDiagram from '/img/gitbook-import/assets/image (3) (1) (2
 
 # How To Implement Dependency Injection
 
-Content in preparation.
+:::info
+For views you can make a simple [ViewLocator](../../concepts/view-locator) which will automatically resolve the view from the Locator.
+:::
 
 This guide will show you how to use Dependency Injection (DI) with _Avalonia UI_ and the MVVM pattern. 
 
 <img src={MvvmArchitectureDiagram} alt=""/>
 
+Let's assume that your MainViewModel as a dependency on MyService. The contructor for ViewModel should looks like this:
+```csharp
+private readonly MyService _myService;
+
+public MapsViewModel(MyService myService)
+{
+    _myService = myService;
+}
+```
+These are the steps you need to do in order to resolve that dependencu using DI.
+
+## Step 1: Install the NuGet package Microsoft.Extensions.DependencyInjection
+```shell
+dotnet add package Microsoft.Extensions.DependencyInjection
+```
+
+## Step 2: ServiceCollectionExtensions 
+The following code is creating an extension for IServiceCollection. This is where you will register all your common dependencies. It's recomended to create that class in the shared project used by all avalonia target platform.
+
+```csharp
+public static class ServiceCollectionExtensions {
+    public static void AddCommonServices(this IServiceCollection collection) {
+        collection.AddTransient<MainViewModel>();
+        collection.AddTransient<MyService>();
+    }
+}
+```
+
+## Step 3: ServiceManager
+The following code create a ServiceManager class that is used to retrieve an instance of a given object.
+```csharp
+public static class ServiceManager {
+    public static IServiceProvider Locator { get; private set; }
+
+    public static void InitialiseLocatorFromCollection(IServiceCollection collection) {
+        Locator = collection.BuildServiceProvider();
+    }
+}
+``` 
+
+## Step 4: Modify Program.cs
+This should be modified for each target platform project. You can add specific platform dependency at this location too (see comment in code below)
+
+```csharp
+class Program
+{
+    // Initialization code. Don't use any Avalonia, third-party APIs or any
+    // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
+    // yet and stuff might break.
+    [STAThread]
+    public static void Main(string[] args)
+    {
+        var collection = new ServiceCollection();
```

Comment:
Please read the follow up comments in the discussion you copied this all from, it was discussed not to create the container like this since it can cause issues.

https://github.com/AvaloniaUI/Avalonia/discussions/13455

Reply:
Yes, Main method is not a universal entry point of the Avalonia app. 
It won't work with Designer (its entry point is BuildAvaloniaApp method). It also won't work with mobile - each platform has their own, like MainActivity on Android. 

Universal place would be Application class, which is always created. Coincidentally, it's a convention place to create a MainWindow, which is usually also a root where IoC is resolved.

Reply:
That would be very helpful, as I still have it in the main, since it works and I currently have other priorities.

Reply:
I've tried to modify the example according the discussion and use AfterSetup(). It's working on my small app but can you check this is correct? Should I also add a comment explaning what you just said above @maxkatz6 ?